src/app/new-page/component/content-flow-item/content-flow-item.component.ts

```diff
@@ -0,0 +1,62 @@
+import { Component, Input, OnChanges, SimpleChanges } from '@angular/core';
+import {
+  Content,
+  FlowItem,
+  FlowWatcher
+} from 'src/app/services/xml-parser-service/xmp-parser.service';
+import { PageService } from '../../service/page-service.service';
+
+@Component({
+  selector: 'app-content-new-flow-item',
+  templateUrl: './content-flow-item.component.html',
+  styleUrls: ['./content-flow-item.component.css']
+})
+export class ContentFlowItemNewComponent implements OnChanges {
+  // eslint-disable-next-line @angular-eslint/no-input-rename
+  @Input() item: FlowItem;
+  contents: Content[];
+  ready: boolean;
+  state: any;
+  isHidden: boolean;
+  flowWatcher: FlowWatcher;
+
+  constructor(private pageService: PageService) {
+    this.state = this.pageService.parserState();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    for (const propName in changes) {
+      if (changes.hasOwnProperty(propName)) {
+        switch (propName) {
+          case 'item': {
+            if (
+              !changes['item'].previousValue ||
+              changes['item'].currentValue !== changes['item'].previousValue
+            ) {
+              this.ready = false;
+              this.init();
+            }
+          }
+        }
+      }
+    }
+  }
+
+  ngOnDestroy(): void {
+    this.flowWatcher.close();
+  }
+
+  private init(): void {
```

Comment:
can this `init()` method be called multiple times? I see the trigger is from `ngOnChanges`, which I'm assuming could be triggered multiple times.

if it is triggered multiple times you will probably want to cancel any previous active watchers before starting your new watcher.

The potential issue with not canceling a stale watcher is that it will continue to be called when the data model changes, and this will leak memory as well as potentially corrupt some UI if the stale watcher triggers a callback with some outdated/invalid state.

Reply:
Gotcha, I've amened the code to ensure we close out watchers before setting up a new one.